File: data-access-layer/modules/property-management/repo.ts
import {
  newPagedResponse,
  type PagedResponse,
} from "../../../types/pagination.ts";
import type {
  Property,
  PropertyChild,
  PropertyParentDocumentRequirement,
  PropertyUser,
} from "./model.ts";
import { Pool } from "pg";
import {
  calculateOffset,
  executeQuery,
  invalidateCache,
  withCacheAsideRedis,
  withWriteThroughRedisCache,
} from "../../shared/util/query.ts";
import type { RedisClientType } from "../../db/redis-client.ts";
import type { DocumentType } from "../../shared/types/reporting.ts";

export interface IPropertyManagementRepo {
  getAllProperties(
    pageSize: number,
    pageNumber: number,
  ): Promise<{ data?: PagedResponse<Property>; error?: Error }>;
  getPropertyUser(
    propertyId: string,
    userId: string,
  ): Promise<{ data?: PropertyUser; error?: Error }>;
  getAllPropertyParentDocumentRequirements(
    propertyId: string,
  ): Promise<{ data?: PropertyParentDocumentRequirement[]; error?: Error }>;
  getAllPropertyChildren(
    propertyId: string,
  ): Promise<{ data?: PropertyChild[]; error?: Error }>;
  incrementPropertyChildrenPointsForGivenParent(
    propertyId: string,
    parentId: string,
    childrenIds: string[],
    pointValue: number,
  ): Promise<{ data?: PropertyChild[]; error?: Error }>;
  getPointValueForGivenPropertyParentDocumentByDocumentType(
    propertyId: string,
    documentType: DocumentType,
  ): Promise<{ data?: number; error?: Error }>;
}

export class PropertyManagementRepo implements IPropertyManagementRepo {
  private pool: Pool;
  private redisClient: RedisClientType;
  constructor(pool: Pool, redisClient: RedisClientType) {
    this.pool = pool;
    this.redisClient = redisClient;
  }

  async getAllProperties(
    pageSize: number,
    pageNumber: number,
  ): Promise<{ data?: PagedResponse<Property>; error?: Error }> {
    const sql = `
    SELECT 
      *,
      COUNT(*) OVER() as total_count
    FROM property_management.properties
    LIMIT $1
    OFFSET $2;
    `;
    const { data, error } = await executeQuery<
      Property & { total_count: number }
    >(this.pool, sql, [pageSize, calculateOffset(pageSize, pageNumber)]);
    if (error) return { data: undefined, error: error };
    const total_count = data?.rows[0].total_count;
    return {
      data: newPagedResponse(
        data?.rows || [],
        total_count || 0,
        pageNumber,
        pageSize,
      ),
      error: undefined,
    };
  }

  async getPropertyUser(
    propertyId: string,
    userId: string,
  ): Promise<{ data?: PropertyUser; error?: Error }> {
    const sql = `
    SELECT *
    FROM property_management.property_users
    WHERE property_id = $1 AND user_id = $2;
    `;
    const { data, error } = await executeQuery<PropertyUser>(this.pool, sql, [
      propertyId,
      userId,
    ]);
    if (error) return { data: undefined, error: error };
    return { data: data?.rows[0], error: undefined };
  }

  async getAllPropertyParentDocumentRequirements(
    propertyId: string,
  ): Promise<{ data?: PropertyParentDocumentRequirement[]; error?: Error }> {
    const sql = `
    SELECT *
    FROM property_management.property_parent_document_requirements
    WHERE property_id = $1;
    `;
    const { data, error } =
      await executeQuery<PropertyParentDocumentRequirement>(this.pool, sql, [
        propertyId,
      ]);
    if (error) return { data: undefined, error: error };
    return { data: data?.rows, error: undefined };
  }

  async getAllPropertyChildren(
    propertyId: string,
  ): Promise<{ data?: PropertyChild[]; error?: Error }> {
    const cacheKey = this.getAllPropertyChildrenCacheKey(propertyId);
    return await withCacheAsideRedis(this.redisClient, cacheKey, async () => {
      const sql = `
        SELECT *
        FROM property_management.property_children
        WHERE property_id = $1;
        `;
      const { data, error } = await executeQuery<PropertyChild>(
        this.pool,
        sql,
        [propertyId],
      );
      if (error) return { data: undefined, error: error };
      return { data: data?.rows, error: undefined };
    });
  }

  async incrementPropertyChildrenPointsForGivenParent(
    propertyId: string,
    parentId: string,
    childrenIds: string[],
    pointValue: number,
  ): Promise<{ data?: PropertyChild[]; error?: Error }> {
    const cacheKey = `properties:${propertyId}:parents:${parentId}:children`;
    const { data, error } = await withWriteThroughRedisCache(
      this.redisClient,
      cacheKey,
      async () => {
        const sql = `
          UPDATE property_management.property_children
          SET points = points + $1
          WHERE property_id = $2
          AND child_id = ANY($3::uuid[])
          RETURNING *;
        `;
        const { data, error } = await executeQuery<PropertyChild>(
          this.pool,
          sql,
          [pointValue, propertyId, childrenIds],
        );
        if (error) return { data: undefined, error };
        return { data: data?.rows, error: undefined };
      },
    );
    if (error) return { data: undefined, error };
    await invalidateCache(
      this.redisClient,
      this.getAllPropertyChildrenCacheKey(propertyId),
    );
    return { data: data, error: undefined };
  }

  async getPointValueForGivenPropertyParentDocumentByDocumentType(
    propertyId: string,
    documentType: DocumentType,
  ): Promise<{ data?: number; error?: Error }> {
    //TODO: invalidate (if necessary)
    const cacheKey = `properties:${propertyId}:parents:documents:${documentType}:points`;
    return await withCacheAsideRedis(this.redisClient, cacheKey, async () => {
      const sql = `
        SELECT point_value
        FROM property_management.property_parent_document_requirements
        WHERE property_id = $1 AND document_type = $2;
        `;
      const { data, error } = await executeQuery<{ point_value: number }>(
        this.pool,
        sql,
        [propertyId, documentType], // TODO: SQL index on document_type
      );
      if (error) return { data: undefined, error };
      return { data: data?.rows[0].point_value, error: undefined };
    });
  }

  private getAllPropertyChildrenCacheKey(propertyId: string): string {
    return `properties:${propertyId}:children`;
  }
}

File: apps/backend/src/modules/property_management/repo.py
from abc import ABC, abstractmethod
from typing import List
from uuid import UUID
import asyncpg
from src.shared.exceptions.not_found import NotFoundException
from src.shared.exceptions.database import DatabaseException
from src.shared.types.modules.property_management.model import (
    Property,
    PropertyChild,
    PropertyChildDocumentRequirement,
    PropertyParentDocumentRequirement,
)
from src.shared.types.pagination import PagedResponse
from src.shared.utils.pagination import calculate_offset, new_paged_response
from src.shared.utils.query import try_except


class IPropertyManagementRepo(ABC):
    @abstractmethod
    async def get_all_properties(
        self,
        page_size: int,
        page_number: int,
    ) -> PagedResponse[Property]:
        pass

    @abstractmethod
    async def get_all_property_parent_document_requirements(
        self,
        property_id: UUID,
    ) -> List[PropertyParentDocumentRequirement]:
        pass

    @abstractmethod
    async def get_all_property_children(
        self,
        property_id: UUID,
    ) -> List[PropertyChild]:
        pass

    @abstractmethod
    async def get_all_property_children_document_requirements(
        self,
        property_id: UUID,
    ) -> List[PropertyChildDocumentRequirement]:
        pass

    @abstractmethod
    async def get_all_property_children_paged(
        self,
        property_id: UUID,
        page_size: int,
        page_number: int,
    ) -> PagedResponse[PropertyChild]:
        pass

    @abstractmethod
    async def increment_property_children_points_for_given_parent(
        self,
        property_id: UUID,
        children_ids: List[UUID],
        point_value: int,
    ) -> List[PropertyChild]:
        pass

    @abstractmethod
    async def get_property_child_by_id(
        self,
        property_id: UUID,
        child_id: UUID,
    ) -> PropertyChild:
        pass

    @abstractmethod
    async def get_all_property_children_for_given_parent(
        self, property_id: UUID, parent_id: UUID
    ) -> List[PropertyChild]:
        pass


class PropertyManagementRepo(IPropertyManagementRepo):
    def __init__(self, pool: asyncpg.Pool) -> None:
        self.pool = pool

    async def get_all_properties(
        self,
        page_size: int,
        page_number: int,
    ) -> PagedResponse[Property]:
        sql = """
        SELECT 
          *,
          COUNT(*) OVER() as total_count
        FROM property_management.properties
        LIMIT $1
        OFFSET $2;
        """
        async with self.pool.acquire() as connection:
            rows, error = await try_except(
                connection.fetch,
                sql,
                page_size,
                calculate_offset(page_size, page_number),
            )
            if error:
                raise DatabaseException(message=str(error))
            if not rows:
                return new_paged_response(
                    items=[],
                    total=0,
                    page_number=page_number,
                    page_size=page_size,
                )

            total_count = rows[0]["total_count"]
            properties = [Property(**row) for row in rows]
            return new_paged_response(
                items=properties,
                total=total_count,
                page_number=page_number,
                page_size=page_size,
            )

    async def get_all_property_parent_document_requirements(
        self,
        property_id: UUID,
    ) -> List[PropertyParentDocumentRequirement]:
        sql = """
          SELECT *
          FROM property_management.property_parent_document_requirements
          WHERE property_id = $1;
          """
        async with self.pool.acquire() as connection:
            rows, error = await try_except(connection.fetch, sql, property_id)
            if error:
                raise DatabaseException(message=str(error))
            if rows is None or len(rows) == 0:
                raise NotFoundException()
            requirements = [
                PropertyParentDocumentRequirement(**dict(row)) for row in rows
            ]
            return requirements

    async def get_all_property_children(
        self,
        property_id: UUID,
    ) -> List[PropertyChild]:
        sql = """
        SELECT *
        FROM property_management.property_children
        WHERE property_id = $1;
        """
        async with self.pool.acquire() as connection:
            rows, error = await try_except(connection.fetch, sql, property_id)
            if error:
                raise DatabaseException(message=str(error))
            if rows is None or len(rows) == 0:
                raise NotFoundException()
            return [PropertyChild(**row) for row in rows]

    async def get_all_property_children_document_requirements(
        self,
        property_id: UUID,
    ) -> List[PropertyChildDocumentRequirement]:
        sql = """
        SELECT *
        FROM property_management.property_children_document_requirements
        WHERE property_id = $1;
        """
        async with self.pool.acquire() as connection:
            rows, error = await try_except(connection.fetch, sql, property_id)
            if error:
                raise DatabaseException(message=str(error))
            if rows is None or len(rows) == 0:
                raise NotFoundException()
            return [PropertyChildDocumentRequirement(**row) for row in rows]

    async def get_all_property_children_paged(
        self,
        property_id: UUID,
        page_size: int,
        page_number: int,
    ) -> PagedResponse[PropertyChild]:
        sql = """
        SELECT 
          *,
          COUNT(*) OVER() as total_count
        FROM property_management.property_children
        WHERE property_id = $1
        LIMIT $2
        OFFSET $3;
        """
        async with self.pool.acquire() as connection:
            rows, error = await try_except(
                connection.fetch,
                sql,
                property_id,
                page_size,
                calculate_offset(page_size=page_size, page_number=page_size),
            )
            if error:
                raise DatabaseException(message=str(error))
            if rows is None or len(rows) == 0:
                return new_paged_response(
                    [], 0, page_number=page_number, page_size=page_size
                )
            total_count = rows[0]["total_count"]
            return new_paged_response(
                items=[PropertyChild(**row) for row in rows],
                total=total_count,
                page_size=page_size,
                page_number=page_number,
            )

    async def increment_property_children_points_for_given_parent(
        self,
        property_id: UUID,
        children_ids: List[UUID],
        point_value: int,
    ) -> List[PropertyChild]:
        sql = """
        UPDATE property_management.property_children
        SET points = points + $1
        WHERE property_id = $2
        AND child_id = ANY($3::uuid[])
        RETURNING *;
        """
        async with self.pool.acquire() as connection:
            rows, error = await try_except(
                connection.fetch,
                sql,
                point_value,
                property_id,
                children_ids,
            )
            if error:
                raise DatabaseException(message=str(error))
            if rows is None or len(rows) == 0:
                raise NotFoundException()
            return [PropertyChild(**row) for row in rows]

    async def get_property_child_by_id(
        self,
        property_id: UUID,
        child_id: UUID,
    ) -> PropertyChild:
        sql = """
        SELECT *
        FROM property_management.property_children
        WHERE property_id = $1 AND child_id = $2;
        """
        async with self.pool.acquire() as connection:
            rows, error = await try_except(connection.fetch, sql, property_id, child_id)
            if error:
                raise DatabaseException(message=str(error))
            if rows is None or len(rows) == 0:
                raise NotFoundException(
                    message=f"Child with id {child_id} is not registered to property: {property_id}!",
                )
            return PropertyChild(**rows[0])

    async def get_all_property_children_for_given_parent(
        self, property_id: UUID, parent_id: UUID
    ) -> List[PropertyChild]:
        sql = """
        SELECT *
        FROM property_management.property_children
        WHERE property_id = $1 AND parent_id = $2;
        """
        async with self.pool.acquire() as connection:
            rows, error = await try_except(
                connection.fetch, sql, property_id, parent_id
            )
            if error:
                raise DatabaseException(message=str(error))
            if rows is None or len(rows) == 0:
                raise NotFoundException()
            return [PropertyChild(**row) for row in rows]
